File: backtracking/211.py
class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.wordarr = {}
        

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        if len(word) in self.wordarr:
            self.wordarr[len(word)].append(word)
        else:
            self.wordarr[len(word)] = [word]
        

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        
        if not self.wordarr or not word:
            return False
        if len(word) not in self.wordarr:
            return False
        
        if not self.wordarr[len(word)]:
            return False
        
        for w in list(self.wordarr[len(word)]):
            count = 0
            for i in range(len(word)):
                if word[i] == '.' or word[i] == w[i]:
                    count+= 1
            if count == len(word):
                return True
        return False
            
    


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: trie/211.go
type WordDictionary struct {
    next map[rune]*WordDictionary
    isWord bool
}


/** Initialize your data structure here. */
func Constructor() WordDictionary {
    return WordDictionary{next: make(map[rune]*WordDictionary), isWord: false}
}


/** Adds a word into the data structure. */
func (this *WordDictionary) AddWord(word string)  {
    for _, c := range word {
        if _, ok := this.next[c]; !ok {
            this.next[c] = &WordDictionary{next: make(map[rune]*WordDictionary), isWord: false}
        }
        this = this.next[c]
    }
    this.isWord = true
}


/** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
func (this *WordDictionary) Search(word string) bool {
    if len(word) == 1 {
        if word[0] == '.' {
            for _, v := range this.next {
                if v.isWord {
                    return true
                }
            }
        }else {
            if _, ok := this.next[rune(word[0])]; !ok {
                return false
            }else {
                return this.next[rune(word[0])].isWord
            }    
        }
    }else {
        if word[0] == '.' {
            for _, v := range this.next {
                if v.Search(word[1:]) {
                    return true
                }
            }
        }else {
            if _, ok := this.next[rune(word[0])]; !ok {
                return false
            }else {
                return this.next[rune(word[0])].Search(word[1:])
            }            
        }
    }
    return false
}


/**
 * Your WordDictionary object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddWord(word);
 * param_2 := obj.Search(word);
 */
